### PR TITLE
Fix dyno resolution of user-defined initializer for owned classes

### PR DESCRIPTION
**Pull request by riftEmber**
Fix a bug causing Dyno to ignore a user-defined initializer when resolving the initializer for an owned class, and a related bug causing ambiguity between parent and child initializers.

When resolving an `init`, dyno discards potential user-defined functions that require conversion to be applicable, including this case since implicit `owned`->`borrowed` conversion is needed for the receiver of `init`. This leads the compiler to determine a compiler-generated `init` is necessary, and the bug later manifests as ambiguity between the user-specified and compiler-generated `init`s. Fixed by treating implicit borrowing as a separate kind of conversion, and making an exception for it to not be discarded.

Additionally, fixed disambiguation between parent and child class `init`s by discarding options that implicitly convert, except those that just do a borrowing conversion per the new conversion kind. Previously this was not possible as discarding options that do _any_ conversion would discard the desired user-defined `init`.

Adds a test locking in the correct behavior for both these bugs. Verified the single test case catches if either bug is left unfixed.

Resolves https://github.com/Cray/chapel-private/issues/5557.

Future work:
- The compiler (prod and dyno) considers `owned C` a subtype of `borrowed C`, per the `is_subtype` primitive or `isSubtype` in the `Type` module. With this PR a distinction is made between a subtype and an implicitly borrowed-from type in dyno, but the behavior of the primitive is preserved for now. This should be fixed in the future; see https://github.com/Cray/chapel-private/issues/5851.

[reviewer info placeholder]

Testing:
- [x] dyno tests
- [x] paratest

### CODE REVIEW

File: frontend/lib/resolution/default-functions.cpp [hunk 1]
```diff
@@ -97,7 +97,11 @@ areOverloadsPresentInDefiningScope(Context* context, const Type* type,
           // if the receiver type is a generic type and we have
           // an instantiation.
           auto result = canPass(context, haveQt, receiverQualType);
-          if (result.passes() && !result.converts() && !result.promotes()) {
+          if (result.passes() &&
+              (!result.converts() ||
```
Comment (mppf) [inline frontend/lib/resolution/default-functions.cpp:101]: Should we have a sibling method to `converts()` that checks for borrowing conversion only?

**Review by mppf — APPROVED** [review]
Thanks for fixing the issue!

For the PR message

> Dyno discards potential user-defined functions that require conversion to be applicable, including this case since implicit

Could you adjust this sentence to be only focused on `init` calls? E.g.


> When resolving an `init`, the dyno resolver discards potential user-defined functions that require conversion to be applicable, including this case since implicit

File: frontend/include/chpl/resolution/can-pass.h [hunk 2]
```diff
@@ -52,6 +52,8 @@ class CanPassResult {
     NUMERIC,
     /** A conversion that implements subtyping */
     SUBTYPE,
+    /** A conversion that borrows a managed type */
+    BORROWS,
```
[the remark this answers is not in the document]
Comment (riftEmber) [reply frontend/include/chpl/resolution/can-pass.h:56]: I think that we would need to distinguish between them -- the bug fixes in this PR rely on being able to tell when a conversion is only an implicit borrow and not an implicit subtype conversion. I considered making `ConversionKind` a bit-set enum but I think it's not worth the complexity as this is probably the only case of overlap, so I lean towards adding a `BORROWS_SUBTYPE` value.

File: frontend/lib/resolution/can-pass.cpp [hunk 3]
```diff
@@ -655,11 +680,20 @@ CanPassResult CanPassResult::canConvert(Context* context,
   const Type* actualT = actualQT.type();
   const Type* formalT = formalQT.type();
 
+  // can we convert with a borrowing conversion (without any subtyping)?
```
[the remark this answers is not in the document]
Comment (riftEmber) [reply frontend/lib/resolution/can-pass.cpp:683]: Will add `BORROWS_SUBTYPE` as separate from `BORROWS` for this case.

**Review by mppf — APPROVED** [review]
Thanks for addressing my feedback!

**Pull request merged**